ISSUE-970: Make input/output schema and config fields optional for CP

priyank5485 opened the pull request:
ISSUE-970: Make input/output schema and config fields optional for CP

priyank5485 commented:
@arunmahadevan  Can you please review the back end changes when you get a chance?

arunmahadevan commented on streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/processor/CustomProcessorInfo.java:
```diff
@@ -222,7 +222,7 @@ private ComponentUISpecification getCustomProcessorUISpecification (ComponentUIS
         }
         result.setFields(fields);
-        return result;
+        return (result.getFields().size() == 0 ? null : result);
```
nit: use fields.isEmpty()
Q: will it cause NPE at the caller? maybe better to return an Optional.

priyank5485 replied:
Changed to isEmpty.
No, it wont cause a NPE. Its a private method and the only call is in this class that relies on it returning null.

arunmahadevan commented on streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/CustomProcessorBoltFluxComponent.java:
```diff
@@ -39,14 +39,13 @@
         Object[] values = new Object[configMethodNames.length];
         values[0] = conf.get(TopologyLayoutConstants.JSON_KEY_CUSTOM_PROCESSOR_IMPL);
         try {
             values[1] = getOutputSchemaJson();
```
getInputSchemaJson is removed but not getOutputSchemaJson but we want to make both optional ?

priyank5485 replied:
Those were config methods to the bolt. InputSchema to the bolt before was never used. So it was an unused instance variable in the bolt. Moreover, it did not make sense to pass it since its provided during CP registration initially. Hence removed that variable and the corresponding method. OutputSchema here is the output schema from the bolt configured by user in SAM app using UI. We need it to declare fields from the CustomProcessorBolt.

arunmahadevan commented on streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/CustomProcessorBolt.java:
```diff
@@ -203,30 +170,35 @@ public void prepare (Map stormConf, TopologyContext context, OutputCollector col
-                if (event.get(entry.getValue()) != null) {
-                    mappedEvent.put(entry.getKey(), event.get(entry.getValue()));
+            StreamlineEvent toProcess = null;
+            if (inputSchemaMap == null || inputSchemaMap.isEmpty() || !inputSchemaMap.containsKey(input.getSourceStreamId())) {
```
1. Can it ever be empty or null? This can be a private final initialized to an empty map.
2. use get to retrieve the mapping based on the stream id before the `if` and check for null and reuse it in the else.

priyank5485 replied:
Yes, it can be empty or null. In case of a CP registered with optional input schema, UI will not pass any mapping and hence it will be empty or null. In this case we pass the event received in the CustomProcessorBolt as is to the CP implementation. Your suggestion is better in general, but I think in this case it will make the code a bit complicated. If you are okay, we can leave it as is.

arunmahadevan commented on streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/CustomProcessorBolt.java:
```diff
@@ -203,30 +170,35 @@ public void prepare (Map stormConf, TopologyContext context, OutputCollector col
-                    .getId(), event.getHeader(), input.getSourceStreamId(), event.getAuxiliaryFieldsAndValues()));
+            List<StreamlineEvent> results = customProcessorRuntime.process(toProcess);
             if (results != null) {
+                Schema schema = outputSchema.values().iterator().next();
```
1. Can there be only one output stream and schema? Is that going to be addressed separately?
2. Isn;t the output schema going to be optional as a part of this PR? if so this may throw an Exception.

priyank5485 replied:
Yes, its only one output schema and stream for now. It can be addressed separately based on the need for that use case. I did not plan to address it immediately.
Output schema is optional for registering the CP. This is the output schema at the component level in a SAM app. UI will force the user to create an output schema and hence this will have exactly one entry.

priyank5485 commented:
@arunmahadevan Thanks for the review. Sorry for delayed response on review. For some reason, I dont get notifications if people comment on my PR. And yes, we are making input schema, output schema and config fields optional while registering a CP. Config fields was already optional but it still had the red asterisk showing in UI. We will just remove that with this PR.

priyank5485 commented:
@arunmahadevan I have pushed a commit addressing your review comments and UI changes for this. Let me know if you think we need any more changes for the back end. At this point, I just need to test this out on a cluster

priyank5485 commented:
@arunmahadevan I have rebased and tested this on a cluster. Looks good. Please review when you get a chance.

arunmahadevan commented:
+1

arunmahadevan commented:
Thanks @priyank5485 , merged to master and HDF-3.1.0.0.  You can close the PR and the related ISSUE.
